### DISCUSSION TITLE
Best way to format source generated code?

### DISCUSSION THREAD

**Discussion by thomhurst** [dotnet/roslyn#75722] [Q&A] [answered]
What's the recommendation on nicely formatting generated code?

At the moment I'm just using a string builder and trying to handle it manually but with line breaks and tabs it gets difficult if the code becomes nested.

I know you can parse a syntax tree and call normalize whitespace, but is that going to dramatically slow down my source generator? I don't want a laggy experience for users.

Thanks

**Comment by CyrusNajmabadi** [comment, marked as answer]
Best way is to just emit in a properly formatted fashion to begin with. Just use a string building abstraction that supports indentation. Indent when you enter a block, and dedent when you leave.

**Reply by thomhurst** [reply]
Yeah that's what I've been doing but appending multi line strings loses the indentation. I'll just have to append a line at a time I guess

**Reply by CyrusNajmabadi** [reply]
> Yeah that's what I've been doing but appending multi line strings loses the indentation.

Update your builder to support that concept.  I have one that just breaks up the input into constituent lines and appends them all, each with the appropriate indentation.

**Reply by thomhurst** [reply]
Will splitting the string result in unnecessary allocations and slow things down though?

**Reply by CyrusNajmabadi** [reply]
you don't have to literally split into strings.  for exampe, you can just use `ROS<char>` to represent the lines.

**Comment by Pentadome** [comment]
When  I played with source generators, I used the [csharpier](https://csharpier.com/docs/API) api to format the code.
